staging/src/k8s.io/api/apps/v1/types.go

```diff
@@ -114,6 +114,40 @@ type RollingUpdateStatefulSetStrategy struct {
 	Partition *int32 `json:"partition,omitempty" protobuf:"varint,1,opt,name=partition"`
 }
 
+// PersistentVolumeClaimDeletePolicyType is a string enumeration of the policies that will determine
+// when volumes from the VolumeClaimTemplates will be deleted when the controlling StatefulSet is
+// deleted or scaled down.
+type PersistentVolumeClaimDeletePolicyType string
+
+const (
+	// RetainPersistentVolumeClaimDeletePolicyType is the default
+	// PersistentVolumeClaimDeletePolicy and specifies that
+	// PersistentVolumeClaims associated with StatefulSet VolumeClaimTemplates
+	// will not be deleted.
+	RetainPersistentVolumeClaimDeletePolicyType PersistentVolumeClaimDeletePolicyType = "Retain"
+	// DeletePersistentVolumeClaimDeletePolicyType specifies that
+	// PersistentVolumeClaims associated with StatefulSet VolumeClaimTemplates
+	// will be deleted in the scenario specified in
+	// StatefulSetPersistentVolumeClaimPolicy.
+	DeletePersistentVolumeClaimDeletePolicyType PersistentVolumeClaimDeletePolicyType = "Delete"
+)
+
+// StatefulSetPersistentVolumeClaimPolicy describes the policy used for PVCs
+// created from the StatefulSet VolumeClaimTemplates.
+type StatefulSetPersistentVolumeClaimPolicy struct {
+	// OnSetDeletion specifies what happens to PVCs created from StatefulSet
+	// VolumeClaimTemplates when the StatefulSet is deleted. The default policy
+	// of `Retain` causes PVCs to not be affected by StatefulSet deletion. The
+	// `Delete` policy causes those PVCs to be deleted.
+	OnSetDeletion PersistentVolumeClaimDeletePolicyType `json:"onSetDeletion,omitempty" protobuf:"bytes,1,opt,name=onSetDeletion,casttype=PersistentVolumeClaimDeletePolicyType"`
```

Comment:
If we had this policy on other objects, the value `Set` feels like it might change.  Was there a discussion somewhere about the naming and why `OnSetDeletion` was chosen vs `OnDeletion` (general) and `OnPodDeletion` (specific) if we needed it?

Reply:
OnDeletion or even OnPodDeletion is too general. For example, with {OnSetDeletion: Retain, OnScaleDown: Delete}, no PVCs will be deleted when pods are deleted. Similarly in no case are PVCs deleted if a pod is manually deleted, or deleted due to a node going down, etc.

I agree that "OnSetDeletion" is a little clunky, but we couldn't come up with another accurate description. OnStatefulSetDeletion maybe is more accurate, but is verbose. Something generic like OnControllerDeletion or OnControllingResourceDeletion feels very obscure and/or imprecise; it might imply behavior changes if a controllerRef is added.

I think a key point is that the deletion is occurring in a situation/context and not because of a specific resource. OnScaleDown captures that very neatly. We also have the advantage there of scale-down being a tightly defined concept associated with a nice short noun. I don't think there's a similarly nice way to describe the deletion of the stateful set other than "StatefulSet Deletion" (or our shortened "SetDeletion")?

At any rate I'd be happy for any suggestion.

Reply:
> OnDeletion or even OnPodDeletion is too general

Citation needed :)

Generally we go for brevity in the api.  If the context is obvious, we don't say the extra word.  So in a pod spec we don't say "pod" because it's obvious this is about the pod.  In a stateful set, we don't say "pod template" because it's the set template, but we do say "pod management policy" to distinguish from management policy.  We don't say "container" on any of the fields inside pod.spec.container.

So right now the default API consistency answer would be "Deleted" is the correct way, in the context of a set.  An argument I might accept is that "a user might be confused and think this is the policy when the PVC is deleted", but only if a user would reasonably make that mistake.

Adding more words has to have a reason and a justification.  That reason has to override two core principles in naming (don't be redundant, and be consistent) in order to be used.  I probably would say `SetDeleted` is too in the middle - it introduces a new word `set` we don't use on its own anywhere (we never say setSelector, for instance) - so it's not consistent.  So either it has to be `Deletion` or `StatefulSetDeletion`.  Since `StatefulSetDeletion` is redundant, we have to show that the redundancy reduces confusion / increases consistency / improves legibility enough to justify redundancy.

Reply:
If you think it's obvious that "deletion" refers only to stateful set deletion and not pod deletion, that's enough for me.

Reply:
As I'm going through the code making this change, it is starting to feel like a WhenDeleted PersistentVolumeClaimRetentionPolicyType refers to when the PVC is deleted, rather than the set.

But I've been staring at this too long. I'll continue to make the change with WhenDeleted unless I hear otherwise from you.

Reply:
I don't think `whenDeleted: Delete` would suggest that the former is referring to a PVC

Reply:
SGTM, thanks.

> Speaking of, which policy applies when I remove the PVC claims from the spec?

I think none of them should be deleted. That user journey didn't come up in the KEP, but the general theme is that PVCs are not deleted unless there's an explicit policy for it.

I think that's okay. This api would be easily extended with the whenClaimTemplateChanged, as you mentioned. We're not trying to support all possible scenarios, and it's better I think to err on the side of not deleting PVCs.

The implementation currently works by putting ownerrefs from the statefulset or pods, as appropriate. So if the volumeclaimtemplate changed, I guess pods would be deleted and the statefulset controller would restart them?

If the replicas and the volumeclaimtemplate was changed at the same time I think we have to be careful that we don't leak PVCs... so that's a great point, thanks for bringing it up. It does seem to be out of scope for the API review, though.